piet-web/src/text.rs

```diff
@@ -138,45 +138,45 @@ impl TextLayoutBuilder for WebTextLayoutBuilder {
     fn build(self) -> Result<Self::Out, Error> {
         self.ctx.set_font(&self.font.get_font_string());
 
-        let line_metrics =
-            lines::calculate_line_metrics(&self.text, &self.ctx, self.width, self.font.size);
-
-        let widths = line_metrics
-            .iter()
-            .map(|lm| text_width(&self.text[lm.start_offset..lm.end_offset], &self.ctx));
-
-        let width = widths.fold(0.0, |a: f64, b| a.max(b));
-
-        Ok(WebTextLayout {
+        let mut layout = WebTextLayout {
             ctx: self.ctx,
             font: self.font,
             text: self.text,
-            line_metrics,
-            width,
-        })
+            line_metrics: Vec::new(),
+            size: Size::ZERO,
+        };
+
+        layout.update_width(self.width)?;
+        Ok(layout)
     }
 }
 
 impl TextLayout for WebTextLayout {
     fn width(&self) -> f64 {
         // precalculated on textlayout build
-        self.width
+        self.size.width
+    }
+
+    fn size(&self) -> Size {
+        self.size
     }
 
-    // TODO refactor this to use same code as build
     fn update_width(&mut self, new_width: impl Into<Option<f64>>) -> Result<(), Error> {
```

Comment:
Should this be renamed "update_size"?

Reply:
the 'width' here refers to our width *constraint*, not to the width of our final layout (although it will impact that width).

The larger question, perhaps, is whether or not we should allow for both width **and** height constraints.

Reply:
Ah, got it.